### ISSUE TITLE
Usage of RangeFacetHandler and TermListFactory 

### ISSUE THREAD

**Issue by dtosato** [NightOwl888/BoboBrowse.Net#3]
@NightOwl888 
Could you show me an example of the usage RangeFacetHandler that exploits the following constructor? 
```c#
public RangeFacetHandler(string name, TermListFactory termListFactory, IEnumerable<string> predefinedRanges);
```
Is `TermListFactory` still necessary? Since BoboBrowse.Net 3.2.0 accepts only strings I am not able to get advantage of a custom `TermListFactory`.

**Comment by NightOwl888** [comment]
I know the design is kind of confusing, especially the part about supplying a factory for a single field type, when supplying a specialized TermList would probably do. But I wanted to remain true to the API of the original as much as possible, despite its confusing use of data types.

To be clear, BoboBrowse does support data types other than string. However, it requires them to always be stored in the Lucene.Net index as string, even if they are another data type. You need to call an overload that uses TermListFactory when the data type in the Lucene.Net index is a formatted number or date to ensure it is compared as a number or a date, rather than a string. This will give you better performance. It also optionally allows you to change the format from the default way that ToString works to a custom format for display on the UI.

So, for example if you have a field that contains USD and you want it to display as $123.45 in the results, you would first store the field in the index as a numeric formatted string padded with leading zeros to a length at least as long as your longest expected number:

```c#
double price = 123.45;
Field f = new Field("price", price.ToString("00000000.00") , Field.Store.NO, Field.Index.NOT_ANALYZED_NO_NORMS);
f.OmitTermFreqAndPositions = true;
document.Add(f);
```

Then when you specify a facet handler, you would call an overload that accepts a TermListFactory, and specify the format you want in the list of results:

```c#
IFacetHandler priceRange = new RangeFacetHandler(
    "price",
    new PredefinedTermListFactory<double>("c", new CultureInfo("en-US")), 
    new string[] { "[0.00 TO 10.00]", "[10.00 TO 25.00]", "[25.00 TO 50.00]", 
    "[50.00 TO 100.00]", "[100.00 TO *]" });
```
The above example passes "c" as the format string, which is the [.NET standard numeric string](http://msdn.microsoft.com/en-us/library/dwhawy9k%28v=vs.110%29.aspx) for currency.

The data is then output into the BrowseHit.StoredFields as currency format ($123.45) when FetchStoredFields is set to true, not the format that is stored in the Lucene.Net index.

### Formatting Range Strings

In my opinion, there is still a piece missing here - the ability to format the range strings any way you like for display on the UI. Unfortunately, the range string formats are required to be a [specific way](https://github.com/NightOwl888/BoboBrowse.Net/wiki/Prebuilt-Facet-Handler-Types#range-strings) throughout the framework. These formats are meant to be machine readable, but aren't the most user-friendly formats possible.

I think there should be a tool of some sort that allows you to pass a format string (such as "{0:c} - {1:c}") so the format of the range could appear on the UI as "$25.00 - $50.00" instead of "[25.00 TO 50.00]". Doing so, would mean that the UI layer would need to have 2 formats - the human readable one and the BoboBrowse format OR there would need to be another tool that could be used to map the custom format to the BoboBrowse format (taking into account culture sensitivity) when posting the format from the UI back to the business layer.

Unfortunately, I haven't thought of a good way to do this as an add-on component that does not require changes to the framework itself, other than parsing the data out of the format string and reformatting it for UI display. But this is too important to leave out of the release entirely, IMO.

Let me know if you have any ideas how this could be added, preferably as a tool in the Support namespace, which acts as a bridge between the Java world and .NET. Alternatively, send a pull request with a prototype implementation.

**Comment by dtosato** [comment]
@NightOwl888 
I will write you soon about the formatting problem. However for the moment I ask you to take a look to the following test method. It tries to format the output but the result does not match my expectation. Any suggestion? 

```c#
[TestMethod]
public void Test()
{
	// Create index.
	string indexPath = @"c:\temp\test";
	{
		using (Lucene.Net.Store.Directory directory = FSDirectory.Open(new System.IO.DirectoryInfo(indexPath)))
		{
			double price;
			using (IndexWriter modifier = new IndexWriter(directory, new StandardAnalyzer(Lucene.Net.Util.Version.LUCENE_30), true, Lucene.Net.Index.IndexWriter.MaxFieldLength.UNLIMITED))
			{
				{
					price = 10.78942;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 99.222;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 50.2;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 7.34;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 2345.1;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 51.0;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 1000.50;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 999.22;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 898.334;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 91100.50;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}
			}
		}
	}

	// Field.
	string field = "price";

	// Facet.
	double lv1 = 0.0;
	double hv1 = 999.22;
	double lv2 = 999.23;
	double hv2 = 100000;
	string lv1Format = lv1.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
	string hv1Format = hv1.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
	string lv2Format = lv2.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
	string hv2Format = hv2.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
	string facetRange1 = string.Format("[{0} TO {1}]", lv1Format, hv1Format);
	string facetRange2 = string.Format("[{0} TO {1}]", lv2Format, hv2Format);
	IFacetHandler priceFacetHandler = new RangeFacetHandler(field, new PredefinedTermListFactory<double>("c", new CultureInfo("en-US")), new List<string> { facetRange1, facetRange2 });

	ICollection<IFacetHandler> handlerList = new IFacetHandler[] {  
		priceFacetHandler};

	// Lucene index.
	using (Lucene.Net.Store.Directory idx = FSDirectory.Open(new System.IO.DirectoryInfo(indexPath)))
	{
		using (IndexReader reader = IndexReader.Open(idx, true))
		{
			// Bobo reader.
			using (BoboIndexReader boboReader = BoboIndexReader.GetInstance(reader, handlerList))
			{
				// Request.
				BrowseRequest browseRequest = new BrowseRequest();
				browseRequest.Count = 10;
				browseRequest.Offset = 0;
				browseRequest.FetchStoredFields = true;

				// Selection.
				BrowseSelection sel = new BrowseSelection(field);
				browseRequest.AddSelection(sel);

				// Query.
				MatchAllDocsQuery query = new MatchAllDocsQuery();

				// Output.
				FacetSpec spec = new FacetSpec();
				spec.OrderBy = FacetSpec.FacetSortSpec.OrderValueAsc;
				spec.MaxCount = 10;
				browseRequest.SetFacetSpec(field, spec);

				// Browse.
				IBrowsable browser = new BoboBrowser(boboReader);
				BrowseResult result = browser.Browse(browseRequest);

				// Results.
				int totalHits = result.NumHits;
				BrowseHit[] hits = result.Hits;
				IDictionary<String, IFacetAccessible> facetMap = result.FacetMap;
				IFacetAccessible facets = facetMap[field];
				List<BrowseFacet> facetVals = facets.GetFacets().ToList();

				// Check.
				Assert.AreEqual(2, facetVals.Count());
				string[] facetValsStr = new string[] { facetVals[0].ToString(), facetVals[1].ToString() };
				Assert.IsTrue(facetValsStr.Contains("[$0.0 TO $999.22](3)"));
				Assert.IsTrue(facetValsStr.Contains("[$999.23 TO $100000](7)"));
			}
		}
	}
}
```

**Comment by NightOwl888** [comment]
The range string technically has 2 purposes:

1. To specify the begin and end values of the range.
2. To act as a key indicating which range is being selected or updated.

The second purpose requires the range string to be consistent so the UI will be able to inform the business layer what values are selected before the browse, and to be able to update the specific range counts on the UI after the browse.

This means that when you specify a range as `[000000.000 TO 000999.220]` in the constructor of the RangeFacetHandler, it must also be specified with exactly the same string in the BrowseSelection.AddValue() method, and the result will return exactly the same string in the BrowseFacet.Value property corresponding to that facet.

Therefore, you can't expect a currency formatted string out of the result. It is formatted exactly the same way as the input so it can be used as a key to update the count of a specific UI element.

Do note that I made it so it works with any standard numeric format - it first attempts to parse using the invariant culture, and then attempts to parse with the culture specified in the facet handler (defaulting to the culture of the current thread if not provided). Therefore, you could format your range string to be `[$0.00 TO $999.22]` and it will work fine in the en-US culture (and I believe the invariant culture, as well).

That said, I still don't think this is good enough - the format string changing from one culture to another could lead to problems. As I mentioned previously, there should be a separate field (or tool) that allows you to convert between the specific string (that acts as a key) and the formatted string for the UI in the current culture.

**Comment by dtosato** [comment]
I modified the code as you wrote and now the test is ok. However, there is still something wrong to me. In particular, it is not clear why if I specify the upper bound of a one interval to 999.21, it includes 7 results, while specifying 999.22 it includes 8. Observing the indexed values, the number of results should be respectively 6 and 7. I suspect that the problem is due to the document which contains the number 99.222. Is that correct?

Here the code of the test.
If you change the value `hv1` you can see the effect. 

```c#
[TestMethod]
public void Test()
{
	// Create index.
	string indexPath = @"c:\temp\test";
	{
		using (Lucene.Net.Store.Directory directory = FSDirectory.Open(new System.IO.DirectoryInfo(indexPath)))
		{
			double price;
			using (IndexWriter modifier = new IndexWriter(directory, new StandardAnalyzer(Lucene.Net.Util.Version.LUCENE_30), true, Lucene.Net.Index.IndexWriter.MaxFieldLength.UNLIMITED))
			{
				{
					price = 10.78942;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 99.222;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 50.2;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 7.34;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 2345.1;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 51.0;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 1000.50;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 999.22;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 898.334;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					price = 91100.50;
					string priceFormatted = price.ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("price", priceFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}
			}
		}
	}

	// Field.
	string field = "price";

	// Facet.
	double lv1 = 0.0;
	double hv1 = 999.21;
	double lv2 = 999.23;
	double hv2 = 100000;
	string lv1Format = lv1.ToString("c", new CultureInfo("en-US"));
	string hv1Format = hv1.ToString("c", new CultureInfo("en-US"));
	string lv2Format = lv2.ToString("c", new CultureInfo("en-US"));
	string hv2Format = hv2.ToString("c", new CultureInfo("en-US"));
	string facetRange1 = string.Format("[{0} TO {1}]", lv1Format, hv1Format); 
	string facetRange2 = string.Format("[{0} TO {1}]", lv2Format, hv2Format);
	IFacetHandler priceFacetHandler = new RangeFacetHandler(field, new PredefinedTermListFactory<double>("c", new CultureInfo("en-US")), new List<string> { facetRange1, facetRange2 });

	ICollection<IFacetHandler> handlerList = new IFacetHandler[] {  
		priceFacetHandler};

	// Lucene index.
	using (Lucene.Net.Store.Directory idx = FSDirectory.Open(new System.IO.DirectoryInfo(indexPath)))
	{
		using (IndexReader reader = IndexReader.Open(idx, true))
		{
			// Bobo reader.
			using (BoboIndexReader boboReader = BoboIndexReader.GetInstance(reader, handlerList))
			{
				// Request.
				BrowseRequest browseRequest = new BrowseRequest();
				browseRequest.Count = 10;
				browseRequest.Offset = 0;
				browseRequest.FetchStoredFields = true;

				// Selection.
				BrowseSelection sel = new BrowseSelection(field);
				browseRequest.AddSelection(sel);

				// Query.
				MatchAllDocsQuery query = new MatchAllDocsQuery();

				// Output.
				FacetSpec spec = new FacetSpec();
				spec.OrderBy = FacetSpec.FacetSortSpec.OrderValueAsc;
				spec.MaxCount = 10;
				browseRequest.SetFacetSpec(field, spec);

				// Browse.
				IBrowsable browser = new BoboBrowser(boboReader);
				BrowseResult result = browser.Browse(browseRequest);

				// Results.
				int totalHits = result.NumHits;
				BrowseHit[] hits = result.Hits;
				IDictionary<String, IFacetAccessible> facetMap = result.FacetMap;
				IFacetAccessible facets = facetMap[field];
				List<BrowseFacet> facetVals = facets.GetFacets().ToList();

				// Check.
				Assert.AreEqual(2, facetVals.Count());
				string[] facetValsStr = new string[] { facetVals[0].ToString(), facetVals[1].ToString() };
				Assert.IsTrue(facetValsStr.Contains("[$0.00 TO $999.21](7)"));
				Assert.IsTrue(facetValsStr.Contains("[$999.23 TO $100,000.00](3)"));
			}
		}
	}
}
```

**Comment by NightOwl888** [comment]
That is correct. The number does not fall within the configured range. Formatting a field does not automatically round its values. You would need to do that explicitly. 

```c#
price = 99.222;
string priceFormatted = Math.Round(price, 2).ToString("000000.000", System.Globalization.CultureInfo.InvariantCulture);
```

Alternatively, you could specify the range with an exclusive upper bound to include every value less than 99.23.

```
[$0.00 TO $99.23)
```

**Comment by dtosato** [comment]
Ok, but `99.23` should follow into the `[$0.0 TO $999.22]` since `99.*` is much less than `999.*`.
This problem seems to be related to the fact that numbers are represented by strings, right?

**Comment by NightOwl888** [comment]
Oops, that was a typo. I updated my last post.

**Comment by dtosato** [comment]
I see, however my issue is still valid.

**Comment by NightOwl888** [comment]
You are right. This appears to be a bug. Oddly, it returns the correct results when you do:

```c#
sel.AddValue(facetRange1);
```
But with no selections at all, it returns 1 number higher than the expected number.

**Comment by NightOwl888** [comment]
This problem existed in the Java version, so I have incremented the version number to 3.2.1-beta00001 to indicate it is patched. I am keeping a 3.2.0 version for parity with the Java code to track fixes in the port separately from patches to the Java version.

I am keeping this issue open because it should be reported to the BoboBrowse team.

Ignore what I told you about being able to format the range strings. This causes the sorting of the ranges to be incorrect. For example, "[$222.22 TO $333.33]" would always come first before "[$99.99 TO $222.21]" when you specify sort in ascending order. You should always use numbers with padded leading zeros to ensure the order is correct ("[0000222.22 TO 0000333.33]"). Again, I think there should be a tool to create a formatted string for display on the UI, but the range strings must be formatted a specific culture-insensitive way in order for all features to work right.

**Comment by NightOwl888** [comment]
FYI - I released a new build with the new [RangeStringFormatter](https://github.com/NightOwl888/BoboBrowse.Net/wiki/Prebuilt-Facet-Handler-Types#custom-range-string-formatting-for-display) functionality. Now it is not very difficult to reformat the range strings for human consumption on the UI. It allows you to specify alternate strings in the case where you have open-ended bounds, so you can make them even more user friendly.

Are you still testing? I would like to try to make this release official so it can be used in production projects.

**Comment by dtosato** [comment]
@NightOwl888 
Yes, I am still here, but I am currently busy on my main project, I will test your implementation at the end of this week or at the beginning of the next one.

**Comment by NightOwl888** [comment]
No problem. I don't expect you to make a full time job out of it :). But thanks for the help tracking down issues so far.

Just let me know if you find any other problems or let me know when you are satisfied that it works according to the documentation (or the original design for those pieces that are not documented) so I can make the release official.

**Comment by dtosato** [comment]
Apparently the  RangeStringFormatter is not available from the last nuget package.
Could you check?
Thanks.

**Comment by NightOwl888** [comment]
Yes, it is available. But it is in the `BoboBrowse.Net.Support` namespace. I put it there because that is the only namespace whose types don't have a counterpart in Java.

Technically, since it is additional functionality above what is in the Java version it should be in a DLL and NuGet package of its own. However, I consider it to be functionality that is missing from RangeStringFormatter, not something that is intended as an optional add-on. So I compromised and put it into the BoboBrowse.Net library in the Support namespace.

In addition to the [documentation about RangeStringFormatter](https://github.com/NightOwl888/BoboBrowse.Net/wiki/Prebuilt-Facet-Handler-Types#custom-range-string-formatting-for-display), there are also [unit tests](https://github.com/NightOwl888/BoboBrowse.Net/blob/master/src/BoboBrowse.Net/BoboBrowse.Tests/Support/RangeStringFormatterTest.cs).

**Comment by dtosato** [comment]
Maybe I doing domething wrong, but I cannot find the RangeStringFormatter  into BoboBrowse.Net.Support namespace.
Here the proof:
![BoboBrowse.Net.Support](https://dl.dropboxusercontent.com/u/20825744/tmp/bobo1.PNG)

**Comment by NightOwl888** [comment]
Are you sure you have the latest version v3.2.1-beta00002 from NuGet? I installed into a brand new project and [`RangeStringFormatter<T>`](https://github.com/NightOwl888/BoboBrowse.Net/blob/master/src/BoboBrowse.Net/BoboBrowse.Net/Support/RangeStringFormatter.cs) is definitely there.

![rangestringformatter](https://cloud.githubusercontent.com/assets/1538288/6350776/6de96d98-bc69-11e4-870b-0aae277b68f0.jpg)

If you do have the latest BoboBrowse.Net beta version, please inform what version of .NET, NuGet, and Visual Studio you are using so I can investigate.

Hmm...now that I think about it, it would make sense to also provide a non-generic version so the type can be provided at runtime, the same as I did with the PredefinedTermListFactory.

**Comment by dtosato** [comment]
Now I see. BoboBrowse.Net v3.2.1-beta0000**2** is not an update of v3.2.1-beta0000**1**. That was the reason of this misunderstood.

**Comment by NightOwl888** [comment]
Well, the original issue here was fixed, so I am closing it. I attempted to contact the BoboBrowse team via email and via Google Groups to no avail. Their [JIRA system](https://linkedin.jira.com/wiki/display/BOBO/Home) doesn't have an obvious way to sign up for it (it gives me an error that my email address isn't from an allowed domain, but doesn't indicate what an "allowed domain" is).

So, there doesn't seem to be any way to report this issue to the BoboBrowse team unless they happen to read this issue. Fat chance.
